gcsio/src/main/java/com/google/cloud/hadoop/gcsio/StorageProvider.java

```diff
@@ -0,0 +1,195 @@
+      GoogleCloudStorageOptions storageOptions,
+      List<ClientInterceptor> interceptors,
+      ExecutorService pCUExecutorService,
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn)
+      throws IOException {
+    final ImmutableMap<String, String> headers = getUpdatedHeadersWithUserAgent(storageOptions);
+
+    return StorageOptions.grpc()
+        .setAttemptDirectPath(storageOptions.isDirectPathPreferred())
+        .setHeaderProvider(() -> headers)
+        .setGrpcInterceptorProvider(
+            () -> {
+              List<ClientInterceptor> clientInterceptorList = new ArrayList<>();
+              if (interceptors != null && !interceptors.isEmpty()) {
+                clientInterceptorList.addAll(
+                    interceptors.stream().filter(x -> x != null).collect(Collectors.toList()));
+              }
+              if (storageOptions.isTraceLogEnabled()) {
+                clientInterceptorList.add(new GoogleCloudStorageClientGrpcTracingInterceptor());
+              }
+
+              if (downscopedAccessTokenFn != null) {
+                // When downscoping is enabled, we need to set the downscoped token for each
+                // request. In the case of gRPC, the downscoped token will be set from the
+                // Interceptor.
+                clientInterceptorList.add(
+                    new GoogleCloudStorageClientGrpcDownscopingInterceptor(
+                        downscopedAccessTokenFn));
+              }
+
+              clientInterceptorList.add(new GoogleCloudStorageClientGrpcStatisticsInterceptor());
+              return ImmutableList.copyOf(clientInterceptorList);
+            })
+        .setCredentials(
+            credentials != null ? credentials : getNoCredentials(downscopedAccessTokenFn))
+        .setBlobWriteSessionConfig(
+            getSessionConfig(storageOptions.getWriteChannelOptions(), pCUExecutorService))
+        .setProjectId(storageOptions.getProjectId())
+        .build()
+        .getService();
+  }
+
+  private static Credentials getNoCredentials(
+      Function<List<AccessBoundary>, String> downscopedAccessTokenFn) {
+    if (downscopedAccessTokenFn == null) {
+      return null;
+    }
+
+    // Workaround for https://github.com/googleapis/sdk-platform-java/issues/2356. Once this is
+    // fixed, change this to return NoCredentials.getInstance();
+    return GoogleCredentials.create(new AccessToken("", null));
+  }
+
+  private void closeStorage(Storage storage) {
+    try {
+      storage.close();
+    } catch (Exception e) {
+      logger.atWarning().withCause(e).log("Error occurred while closing the storage client");
+    }
+  }
+
+  /**
+   * Signal the storage object to be closed. The resources held by the storage object will be freed
+   * only if the instance is closed by all the objects sharing this storage reference.
+   */
+  void close(Storage storage) {
+    if (!storageClientToReferenceMap.containsKey(storage)) {
```

Comment:
Same as while creating storage object, this method implementation is not thread safe. Use the threadsafe modification interfaces of ConcurrentHashMap.

Reply:
Done

Reply:
Used synchronized blocks as discussed offline.